### PR TITLE
Don't read from connection on inactive state.

### PR DESCRIPTION
**Pull request by tcard**
After the connection gets to an inactive state (ie. not CONNECTING,
CONNECTED, CLOSING), no further messages are expected. Instead, moving
back to an active state will start the loop again by calling setConn.

### CODE REVIEW

File: ably/realtime_conn.go [hunk 1]
```diff
@@ -312,6 +312,12 @@ func (c *Conn) isActive() bool {
 	return c.state.current == StateConnConnecting || c.state.current == StateConnConnected
 }
 
+func (c *Conn) lockCanReceiveMessages() bool {
+	c.state.Lock()
+	defer c.state.Unlock()
+	return c.state.current == StateConnConnecting || c.state.current == StateConnConnected || c.state.current == StateConnClosing
```
Comment (gernest) [inline ably/realtime_conn.go:318]: This is  duplicate of `(*Conn) lockIsActive` which calls this
```
func (c *Conn) isActive() bool {
	return c.state.current == StateConnConnecting || c.state.current == StateConnConnected
}
```

So probably adding `|| c.state.current == StateConnClosing` to `(*Conn) isActive` and then calling `lockIsActive()` makes more sense in this situation.

From your pr message you can receive message when in active state so probably `lockIsActive()` alone will do?

File: ably/realtime_conn.go [hunk 2]
```diff
@@ -330,7 +336,7 @@ func (c *Conn) logger() *LoggerOptions {
 func (c *Conn) eventloop() {
 	var receiveTimeout time.Duration
 
-	for {
+	for c.lockCanReceiveMessages() {
```
Comment (gernest) [inline ably/realtime_conn.go:339]: I think `c.lockIsActive()` makes more sense here.

File: ably/realtime_conn.go (hunk 1 above)
Comment (tcard) [reply ably/realtime_conn.go:318]: `isActive` determines whether you can do "active" things (ie. send messages), which you can't if you're closing. But you still can receive messages while closing, hence the different condition.

File: ably/realtime_conn.go (hunk 2 above)
Comment (tcard) [reply ably/realtime_conn.go:339]: See https://github.com/ably/ably-go/pull/149#discussion_r412226818.

File: ably/realtime_conn.go (hunk 1 above)
Comment (gernest) [reply ably/realtime_conn.go:318]: @tcard so  you are describing `c. isActive()||c.state.current == StateConnClosing` ?

Comment (tcard) [reply ably/realtime_conn.go:318]: Yes, that would be equivalent.

**Comment by tcard** [comment]
@gernest Thanks! I'm not sure it does, but anyway, I'd expect this to be seriously refactored as we implement the rest of the recovery logic.

**Pull request merged**